### PR TITLE
revset: bump generation numbers in API to 64 bits

### PR DESCRIPTION
**Pull request by martinvonz**
A chain of 4 billion commits is a lot, but it's not out of the question, so let's support it. The current default index will not be able to handle that many commits, so I let that still use 32-bit integers.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

**Review by PhilipMetzger — COMMENTED** [review]
I'm unsure about the casting, but it's generally fine.

File: lib/src/default_revset_engine.rs [hunk 1]
```diff
@@ -600,7 +600,8 @@ impl<'index> EvaluationContext<'index> {
                 if generation == &GENERATION_RANGE_FULL {
                     Ok(Box::new(RevWalkRevset { walk }))
                 } else {
-                    let walk = walk.filter_by_generation(generation.clone());
+                    let walk =
+                        walk.filter_by_generation(generation.start as u32..generation.end as u32);
```
[the remark this answers is not in the document]
Comment (Ralith) [reply lib/src/default_revset_engine.rs:604]: Isn't `generation.start`/`end` a `u64` here? There's no lossless (`into`) conversion to `u32`.

Would a checked conversion (`try_into`/`try_from`) be better than silently truncating here?

[the remark this answers is not in the document]
Comment (martinvonz) [reply lib/src/default_revset_engine.rs]: Fixed. I went with `unwrap_or()`.

File: lib/src/default_revset_engine.rs [hunk 2]
```diff
@@ -565,6 +566,12 @@ struct EvaluationContext<'index> {
     composite_index: CompositeIndex<'index>,
 }
 
+fn to_u32_generation_range(range: &Range<u64>) -> Range<u32> {
+    let start = range.start.try_into().unwrap();
+    let end = range.end.try_into().unwrap_or(u32::MAX);
```
Comment (yuja) [inline lib/src/default_revset_engine.rs:571]: nit: I would make both start/end do `.unwrap_or(u32::MAX)` assuming there would be no linear history containing `u32::MAX` entries.

If we add a revset function/argument to specify generation, it's possible to specify unreasonably large `start` value.

**Review by yuja — APPROVED** [review]

File: lib/src/default_revset_engine.rs (hunk 2 above)
Comment (martinvonz) [reply lib/src/default_revset_engine.rs:576]: I was thinking that the `unwrap_or()` on the upper bound was basically to handle the default "infinite" range, but that's a good point about future support for letting the user specify the generation range. I made this case error out gracefully instead.

Comment (yuja) [reply lib/src/default_revset_engine.rs:576]: Yes, it's mainly for the infinite range, but also works for future extension. For the lower bound, erroring out should also be good. Thanks.

**Review by yuja — APPROVED** [review]

**Pull request merged**